File: 133_Graph_Clone/cloneGraph.py
class UndirectedGraphNode(object):
    def __init__(self, x):
        self.label = x
        self.neighbors = []

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: UndirectedGraphNode
        :rtype: UndirectedGraphNode
        """
        if not node:
            return None

        root = UndirectedGraphNode(node.label)
        # node.label: clone UndirectedGraphNode
        stack = [node]
        visit = {}
        visit[node.label] = root

        while stack:
            top = stack.pop()
            for n in top.neighbors:
                if n.label not in visit:
                    stack.append(n)
                    visit[n.label] = UndirectedGraphNode(n.label)
                visit[top.label].neighbors.append(visit[n.label])

        return root

File: 133_Graph_Clone/cloneGraph.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

struct UndirectedGraphNode {
    int label;
    vector<UndirectedGraphNode*> neighbors;
    UndirectedGraphNode(int x): label(x) {}
};

class Solution {
    public:
        UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
            if(!node) return NULL;

            auto it = visited.find(node->label);
            if(it != visited.end()) {
                return visited[node->label];
            }

            UndirectedGraphNode* clone = new UndirectedGraphNode(node->label);
            visited[node->label] = clone;

            for(auto neighbor : node->neighbors) {
                (clone->neighbors).push_back(cloneGraph(neighbor));
            }

            return clone;
        }
    private:
        unordered_map<int, UndirectedGraphNode*> visited;
};

int main() {
    return 0;
}
